### PR TITLE
Feature/dictionary

### PR DESCRIPTION
**Pull request by darshan-mehta-zs**
please review the code and the structure.

### CODE REVIEW

File: src/main/java/com/zs/dictionary/Dictionary.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
+package main.java.com.zs.dictionary;
+
+import java.io.*;
+import java.text.MessageFormat;
+import java.util.*;
+import java.util.logging.*;
```
Comment (nikhil-rss-zs) [inline src/main/java/com/zs/dictionary/Dictionary.java]: Fix the auto-imports

File: src/main/java/com/zs/dictionary/Dictionary.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+ * Dictionary to search word from dictionary, to add word to dictionary,
+ * to find words starting with a word or by a difference of one character from dictionary
+ */
+
+public class Dictionary {
+
+    private static Logger logger;
+
+
+    /**
```
Comment (nikhil-rss-zs) [inline src/main/java/com/zs/dictionary/Dictionary.java:26]: Java docs should be more descriptive

File: src/main/java/com/zs/bodmas/Bodmas.java [hunk 3]
```diff
@@ -1,14 +1,22 @@
...
+    /**
+     * @param expression
+     * @return the value after evaluating the expression
+     */
     public int evaluate(String expression) {
 
-        Stack<Integer> numbers = new Stack<>();
-        Stack<Character> operands = new Stack<>();
+        Deque<Integer> numbers = new ArrayDeque<>();
+        Deque<Character> operands = new ArrayDeque<>();
```
Comment (shubhamjsh5175) [inline src/main/java/com/zs/bodmas/Bodmas.java:19]: please check for redundancy in line 25 and 27 @darshan-zopsmart

File: src/main/java/com/zs/bodmas/Bodmas.java [hunk 4]
```diff
@@ -23,21 +31,26 @@ public int evaluate(String expression) {
...
-                    numbers.push(applyOp(operands.pop(), numbers.pop(), numbers.pop()));
+                while (!operands.isEmpty() && hasPrecedence(expression.charAt(i), operands.peek()))
+                    numbers.push(applyOperation(operands.pop(), numbers.pop(), numbers.pop()));
                 operands.push(expression.charAt(i));
             }
         }
-        while (!operands.empty())
-            numbers.push(applyOp(operands.pop(), numbers.pop(), numbers.pop()));
+        while (!operands.isEmpty())
+            numbers.push(applyOperation(operands.pop(), numbers.pop(), numbers.pop()));
```
Comment (shubhamjsh5175) [inline src/main/java/com/zs/bodmas/Bodmas.java:50]: your program is not giving a correct output. Please debug this and find what is going wrong and discuss with me.
Also take care of the boundary conditions like divide by zero and what if you get some unknown character in you string.

File: src/main/java/com/zs/dictionary/Dictionary.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            String line = fr.readLine();
+            while (line != null) {
+                String[] fullLine = line.split("=");
+                String key = fullLine[0];
+                String value = fullLine[1];
+                dictionary.put(key, value);
+                line = fr.readLine();
+            }
+        } catch (FileNotFoundException fe) {
+            logger.severe("File not found");
```
Comment (shubhamjsh5175) [inline src/main/java/com/zs/dictionary/Dictionary.java:123]: Good use of exception handling .

File: src/main/java/com/zs/programs/DecimalToBinary.java [hunk 6]
```diff
@@ -1,14 +1,22 @@
...
+ *
+ */
 public class DecimalToBinary {
 
+    /**
+     * @param decimal
+     * @return binary of the decimal
+     */
     String convertToBinary(int decimal) {
         return Integer.toBinaryString(decimal);
```
Comment (shubhamjsh5175) [inline src/main/java/com/zs/programs/DecimalToBinary.java]: Do not use in-build function as this assignment is given to check your coding skills.

File: src/main/java/com/zs/programs/EmiCalculator.java [hunk 7]
```diff
@@ -23,21 +26,32 @@ public void input() {
         numberOfYears = scanner.nextInt();
     }
```
Comment (shubhamjsh5175) [inline src/main/java/com/zs/programs/EmiCalculator.java:27]: Principle amount can be a non integer.

File: src/main/java/com/zs/programs/ExtractWords.java [hunk 8]
```diff
@@ -1,35 +1,43 @@
...
 public class ExtractWords {
 
+    /**
+     * @param sentence
+     * @return words from the given sentence
+     */
     public String[] getWordsFromSentence(String sentence) {
 
-        //return sentence.split(" ");
         String word = "";
```
Comment (shubhamjsh5175) [inline src/main/java/com/zs/programs/ExtractWords.java:19]: check for logic.

File: src/main/java/com/zs/programs/TwoDimensionArraySum.java [hunk 9]
```diff
@@ -1,14 +1,17 @@
...
+package main.java.com.zs.programs;
 
 import java.util.Scanner;
 
+/**
+ * Program does the sum of two dimensional array
+ */
 public class TwoDimensionArraySum {
 
     int twoD[][];
```
Comment (shubhamjsh5175) [inline src/main/java/com/zs/programs/TwoDimensionArraySum.java]: make this static

File: .gitignore [hunk 10]
```diff
@@ -4,4 +4,4 @@ out/
 *.iml
```
Comment (shubhamjsh5175) [inline .gitignore]: please check why the iml file is included in git

File: src/main/java/com/zs/programs/OneDimensionArraySum.java [hunk 11]
```diff
@@ -20,13 +24,19 @@ public void inputOneD() {
...
             sum += value;
         return sum;
     }
 
+    /**
+     * @param args
+     */
     public static void main(String[] args) {
 
         OneDimensionArraySum oneDimensionArraySum = new OneDimensionArraySum();
```
Comment (shubhamjsh5175) [inline src/main/java/com/zs/programs/OneDimensionArraySum.java:44]: use default/parametrised constructors.

**Review by shubhamjsh5175 — CHANGES_REQUESTED** [review]

**Review by shubhamjsh5175 — APPROVED** [review]

**Review by nikhil-rss-zs — APPROVED** [review]

**Pull request merged**